_posts/2020-09-01-processing-java-stream.md

```diff
@@ -0,0 +1,128 @@
+---
+layout:    post
+title:     "Java Stream - przetwarzanie elementów"
+date:      2020-09-01 08:00:00 +0100
+published: true
+author:    tdrag
+image:     /assets/img/posts/2020-09-01-java-stream-processing/waterfall-stream.jpg
+tags:
+    - java
+    - stream
+    - benchmarking
+---
+
+# Java Stream
+Strumienie zostały dodane do Javy w wersji 8, wzbogacając język o namiastkę programowania funkcyjnego oraz alternatywę dla dobrze znanych pętli.
+Pierwsze pytania jakie się nasuwają, to czy warto stosować strumienie oraz jak robić to świadomie.
+Artykuł ma na celu analizę kilku podstawowych właściwości strumieni oraz porównanie ich z pętlami pod względem wydajności.
+
+# Intermediate vs Terminal
+Java Stream API definiuje dwa rodzaje operacji, jakie możemy wykonać w trakcie przetwarzania strumieni, operacje pośrednie – intermediate,
+oraz operacje końcowe - terminal.
+Operacje pośrednie w sposób deklaratywny opisują, jak dane powinny być przetworzone, a dodatkowo zawsze zwracają obiekt typu **Stream**, co daje możliwość łączenia ich z kolejnymi operacjami pośrednimi.
+Jak widać na przykładzie strumień danych stworzony z listy filtruje te, które zaczynają się na literę "a",
+następnie zamienia wszystkie litery na litery wielkie, a na koniec wyświetla przetworzone elementy.
+Niestety uruchamiając taki kod nic się nie zadzieje, a konsola nie wypisze żadnego elementu
+ponieważ w naszym strumieniu brakuje operacji końcowej -  czyli takiej, która kończy strumień i sprawia, że nie można go już więcej przetwarzać.
+
+```java
+    private static final List<String> LIST = Arrays.asList("aab", "aab", 
+                                                           "aac", "aac", 
+                                                           "bbb", "ccc", "ddd", 
+                                                           "eee", "fff", "ggg");
+    public static void main(String[] args) {
+        LIST.stream()
+                .filter(e -> e.startsWith("a"))
+                .map(String::toUpperCase)
+                .peek(System.out::println);
+
+    }
+```
+
+Po dodaniu do kodu metody collect, dostajemy zamierzony efekt i mimo, że stworzona z przetworzonych danych lista nie jest przypisana do zmiennej, to strumień miał powód aby się wykonać.
+Listę metod pośrednich oraz końcowych znajdziemy [tutaj](https://stackoverflow.com/questions/47688418/what-is-the-difference-between-intermediate-and-terminal-operations).
```

Comment:
nie ma jakiejś bardziej oficjalnej listy (np: w dokumentacji), link do stackoverflow jakoś niezbyt mi się podoba

Reply:
Nie udało mi się znaleźć nic na stronach oracle.
Może coś takiego ?
https://javaconceptoftheday.com/java-8-stream-intermediate-and-terminal-operations/